MapboxCoreNavigation/RouteProgress.swift

```diff
@@ -214,10 +221,46 @@ open class RouteProgress: NSObject {
 
                     congestionTimesPerStep[legIndex].append(stepCongestionValues)
                 }
+
+                if let segmentMaximumSpeedLimits = leg.segmentMaximumSpeedLimits {
+                    guard nextManeuverCoordinateIndex < segmentMaximumSpeedLimits.count else { continue }
+
+                    let stepSegmentMaximumSpeedLimits = Array(segmentMaximumSpeedLimits[maneuverCoordinateIndex..<nextManeuverCoordinateIndex])
+                    maximumSpeedLimitsByStep.append(stepSegmentMaximumSpeedLimits)
+
+                    let polyline = Polyline(coordinates)
+                    let initialCoordinate = coordinates.first
+                    var maximumSpeedPerCoordinate: [(SpeedLimit, CLLocationDistance)] = []
+
+                    for (index, speedLimit) in stepSegmentMaximumSpeedLimits.enumerated() {
+                        let currentCoordinate = coordinates[index]
+                        let distanceTraveled = polyline.distance(from: initialCoordinate, to: currentCoordinate)
```

Comment:
This code is suboptimal because `Polyline.distance(from:to:)` traverses the entire polyline up to the `to` coordinate on every iteration of the for loop. Instead, calculate just the straight-line distance between the two coordinates (using `CLLocationCoordinate2D.distance(to:)` and accumulate the total distance in a local variable outside the for loop. That approach will also avoid inaccuracies when a route doubles back on itself.

Reply:
Good to know about the Polyline.distance algorithm. Switched to the approach you suggested.